src/testing/pod_adaptive_sampling.cpp

```diff
@@ -270,38 +286,44 @@ bool AdaptiveSampling<dim, nstate>::placeTriangulationROMs(const MatrixXd& rom_p
+void AdaptiveSampling<dim, nstate>::updateNearestExistingROMs(const RowVectorXd& /*parameter*/) const{
+
+    pcout << "Verifying ROM points for recomputation." << std::endl;
     //Assemble ROM points in a matrix
     MatrixXd rom_points(0,0);
     for(auto it = rom_locations.begin(); it != rom_locations.end(); ++it){
-        rom_points.conservativeResize(rom_points.rows()+1, 2);
-        rom_points.row(rom_points.rows()-1) = it->first;
+        rom_points.conservativeResize(rom_points.rows()+1, it->get()->parameter.cols());
+        rom_points.row(rom_points.rows()-1) = it->get()->parameter;
     }
 
-    //Get distances between newly added snapshots and ROM points
-    ProperOrthogonalDecomposition::MinMaxScaler scaler;
-    MatrixXd scaled_rom_params = scaler.fit_transform(rom_points);
-    RowVectorXd scaled_point = scaler.transform(parameter);
+    //Get distances between each ROM point and all other ROM points
+    for(auto point : rom_points.rowwise()) {
+        ProperOrthogonalDecomposition::MinMaxScaler scaler;
+        MatrixXd scaled_rom_points = scaler.fit_transform(rom_points);
+        RowVectorXd scaled_point = scaler.transform(point);
 
-    VectorXd distances = (scaled_rom_params.rowwise() - scaled_point).rowwise().squaredNorm();
+        VectorXd distances = (scaled_rom_points.rowwise() - scaled_point).rowwise().squaredNorm();
 
-    std::vector<int> index(distances.size());
-    std::iota(index.begin(), index.end(), 0);
+        std::vector<int> index(distances.size());
+        std::iota(index.begin(), index.end(), 0);
 
-    std::sort(index.begin(), index.end(),
-              [&](const int& a, const int& b) {
-                  return distances[a] < distances[b];
-              });
+        std::sort(index.begin(), index.end(),
+                  [&](const int &a, const int &b) {
+                      return distances[a] < distances[b];
+                  });
 
-    //For 2(n+1) nearest points, if error>tolerance, recompute adjoint 1
-    pcout << "Searching ROM points near: " << parameter << std::endl;
-    for(int i = 0 ; i < 6 ; i++){
-        pcout << "ROM point: " << rom_locations[index[i]].first << " Error: " << rom_locations[index[i]].second->total_error << std::endl;
-        if(std::abs(rom_locations[index[i]].second->total_error) > tolerance){
+        pcout << "Searching ROM points near: " << point << std::endl;
+        double local_mean_error = 0;
+        for (int i = 1; i < rom_points.cols() + 2; i++) {
+            //pcout << "ROM point: " << rom_locations[index[i]]->parameter << " Error: " << rom_locations[index[i]]->total_error << std::endl;
+            local_mean_error = local_mean_error + std::abs(rom_locations[index[i]]->total_error);
+        }
+        local_mean_error = local_mean_error / (rom_points.cols() + 1);
+        if ((std::abs(rom_locations[index[0]]->total_error) > 5 * local_mean_error) || (std::abs(rom_locations[index[0]]->total_error) < 1/5 * local_mean_error)) {
```

Comment:
Should the coefficients 5 and 1/5 be a variable/parameter?

Reply:
I have made this a parameter in the most recent commit.

Reply:
Great! The purpose of the coefficients are much clearer now.